pkg/che/tokencache.go

```diff
@@ -0,0 +1,156 @@
+	if user == "" || pass == "" {
+		return TokenSet{}, fmt.Errorf("the che admin username and/or password are not configured")
+	}
+
+	// get keycloak URL
+	cheKeycloakURL, err := getRouteURL(cl, cfg.GetCheNamespace(), cfg.GetCheKeycloakRouteName())
+	if err != nil {
+		return TokenSet{}, err
+	}
+	log.Info("Che Keycloak Route", "URL", cheKeycloakURL)
+
+	// get che user token
+	if err = tc.obtainNewToken(cheKeycloakURL+tokenPath, user, pass); err != nil {
+		return TokenSet{}, err
+	}
+
+	return *tc.token, nil
+}
+
+// obtainNewToken obtains an access token from the provided authentication URL and returns it, otherwise returns an error
+func (tc *tokenCache) obtainNewToken(authURL, user, pass string) error {
+	reqData := url.Values{}
+	reqData.Set("username", user)
+	reqData.Set("password", pass)
+	reqData.Set("grant_type", "password")
+	reqData.Set("client_id", "che-public")
+
+	log.Info("Obtaining new token", "URL", authURL)
+	res, err := tc.httpClient.PostForm(authURL, reqData)
+	if err != nil {
+		return err
+	}
+
+	defer rest.CloseResponse(res)
+	if res.StatusCode != http.StatusOK {
+		bodyString := rest.ReadBody(res.Body)
+		return errors.Errorf("unable to obtain access token for che, Response status: %s. Response body: %s", res.Status, bodyString)
+	}
+	tokenSet, err := readTokenSet(res)
+	if err != nil {
+		return err
+	}
+	if tokenSet.AccessToken == "" {
+		return errors.New("unable to obtain access token for che. Access Token is missing in the response")
+	}
+	if tokenSet.Expiration == 0 && tokenSet.ExpiresIn > 0 {
+		timeLeft := time.Duration(tokenSet.ExpiresIn) * time.Second
+		tokenSet.Expiration = time.Now().Add(timeLeft).Unix()
+	}
+	log.Info("Token expiry", "expiration", tokenSet.Expiration, "expires_in", tokenSet.ExpiresIn)
+	tc.token = tokenSet
+	return nil
+}
+
+// TokenSet represents a set of Access and Refresh tokens
+type TokenSet struct {
+	AccessToken  string `json:"access_token"`
+	ExpiresIn    int64  `json:"expires_in"`
+	Expiration   int64  `json:"expiration"`
+	RefreshToken string `json:"refresh_token"`
+	TokenType    string `json:"token_type"`
+}
+
+// readTokenSet extracts json with token data from the response
+func readTokenSet(res *http.Response) (*TokenSet, error) {
+	buf := new(bytes.Buffer)
+	_, err := io.Copy(buf, res.Body)
+	if err != nil {
+		return nil, err
+	}
+	jsonString := strings.TrimSpace(buf.String())
+	return readTokenSetFromJSON(jsonString)
+}
+
+// readTokenSetFromJSON parses json with a token set
+func readTokenSetFromJSON(jsonString string) (*TokenSet, error) {
+	var token TokenSet
+	err := json.Unmarshal([]byte(jsonString), &token)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error when unmarshal json with access token %s ", jsonString)
+	}
+	return &token, nil
+}
+
+// tokenExpired return false if the token is not nil and good for at least one more minute
+func tokenExpired(token *TokenSet) bool {
+	return token == nil || time.Now().After(time.Unix(token.Expiration-60, 0))
```

Comment:
What is the default expiration time of the token?

Reply:
5 minutes

Reply:
ok, then I would make the time cushion maybe a bit smaller - eg. 10seconds - or is there any reason why do we need to check that it will be valid at least one minute?